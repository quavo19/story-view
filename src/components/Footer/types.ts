import type React from 'react';
import type {
  ImageProps,
  ImageStyle,
  StyleProp,
  TextInput,
  TextInputProps,
  TextProps,
  TextStyle,
  ViewProps,
  ViewStyle,
} from 'react-native';

export type FooterComponentProps = React.ReactElement<FooterProps>;

export type FooterProps = TextInputProps & {
  label?: string;
  sendIconProps?: ImageProps;
  sendTextProps?: TextProps;
  containerViewProps?: ViewProps;
  shouldShowTextInputSend?: boolean;
  customInput?: TextInput | null;
  shouldShowSendImage?: boolean;
  sendTextStyle?: TextStyle;
  sendIconStyle?: ImageStyle;
  inputStyle?: StyleProp<TextStyle>;
  containerStyle?: ViewStyle;
  sendText?: string;
  onIconPress?: () => void | null;
  onSendTextPress?: () => void | null;
};
