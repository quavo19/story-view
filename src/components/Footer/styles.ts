import { StyleSheet } from 'react-native';
import { Colors } from '../../theme';
import { moderateScale, scale, verticalScale } from '../../theme/Metrics';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    paddingBottom: verticalScale(8),
    alignItems: 'center',
  },
  input: {
    flex: 1,
    paddingVertical: verticalScale(15),
    paddingHorizontal: scale(12),
    color: Colors.white,
    height: verticalScale(45),
    fontSize: moderateScale(12, false, 0.3),
  },
  sendIcon: {
    width: moderateScale(24),
    height: moderateScale(24),
    marginRight: scale(8),
    tintColor: Colors.white,
  },
  sendText: {
    color: Colors.white,
    paddingRight: scale(16),
    fontSize: moderateScale(12, false, 0.3),
  },
  sectionStyle: {
    flexDirection: 'row',
    borderWidth: 1,
    borderColor: Colors.white,
    borderRadius: moderateScale(24),
    overflow: 'hidden',
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    marginHorizontal: scale(8),
  },
  ImageStyle: {
    padding: moderateScale(10),
    margin: moderateScale(5),
    height: moderateScale(25),
    width: moderateScale(25),
    resizeMode: 'stretch',
    alignItems: 'center',
  },
});

export default styles;
