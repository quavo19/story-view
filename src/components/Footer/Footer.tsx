import React, { useRef } from 'react';
import {
  Image,
  StyleSheet,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from 'react-native';
import { Icons } from '../../assets';
import { Strings } from '../../constants';
import { useKeyboardListener } from '../../hooks';
import { Colors } from '../../theme';
import styles from './styles';
import type { FooterProps } from './types';

const Footer = ({
  onIconPress,
  onSendTextPress,
  sendTextStyle,
  sendIconStyle,
  inputStyle,
  containerStyle,
  sendText,
  shouldShowSendImage = true,
  shouldShowTextInputSend = true,
  sendIconProps,
  sendTextProps,
  containerViewProps,
  customInput,
  ...rest
}: FooterProps) => {
  const isKeyboardVisible = useKeyboardListener();
  const ref = useRef<TextInput>(null);

  const handleSendTextPress = () => {
    ref?.current?.clear();
    onSendTextPress?.();
  };

  const _sendTextStyle = StyleSheet.flatten([styles.sendText, sendTextStyle]);
  const _sendIconStyle = StyleSheet.flatten([styles.sendIcon, sendIconStyle]);
  const _inputStyle = StyleSheet.flatten([styles.input, inputStyle]);
  const _containerStyle = StyleSheet.flatten([
    styles.container,
    containerStyle,
  ]);

  return (
    <View style={_containerStyle} {...containerViewProps}>
      <View style={styles.sectionStyle}>
        <>
          {customInput ?? (
            <TextInput
              ref={ref}
              style={_inputStyle}
              placeholder={Strings.sendMessage}
              placeholderTextColor={Colors.white}
              {...rest}
            />
          )}
        </>
        {isKeyboardVisible && shouldShowTextInputSend && (
          <TouchableOpacity onPress={handleSendTextPress}>
            <Text style={_sendTextStyle} {...sendTextProps}>
              {sendText ?? Strings.send}
            </Text>
          </TouchableOpacity>
        )}
      </View>
      {!isKeyboardVisible && shouldShowSendImage && (
        <TouchableOpacity onPress={onIconPress} testID="footerIcon">
          <Image
            source={Icons.send}
            style={_sendIconStyle}
            {...sendIconProps}
          />
        </TouchableOpacity>
      )}
    </View>
  );
};

export default Footer;
