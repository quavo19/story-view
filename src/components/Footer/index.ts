import Footer from './Footer';

export { Footer };
export type { FooterProps } from './types';
