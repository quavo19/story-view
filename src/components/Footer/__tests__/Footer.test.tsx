import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import Footer from '../Footer';

jest.useFakeTimers();

describe('Footer Component', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(<Footer />);
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('Footer TextInput', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(<Footer placeholder="Send Message" />);
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('Footer Icon Props', () => {
  it('Match Snapshot', () => {
    const { toJSON } = render(<Footer shouldShowSendImage={false} />);
    expect(toJSON()).toMatchSnapshot();
  });
  test('onIconPress event', () => {
    const onPressEvent = jest.fn();
    const { getByTestId } = render(<Footer onIconPress={onPressEvent} />);

    const sendIconButton = getByTestId('footerIcon');
    fireEvent.press(sendIconButton);
    expect(onPressEvent).toHaveBeenCalled();
  });
  test('hide send icon', () => {
    const { queryByTestId } = render(<Footer shouldShowSendImage={false} />);
    const sendIconButton = queryByTestId('footerIcon');
    expect(sendIconButton).toBe(null);
  });
});
