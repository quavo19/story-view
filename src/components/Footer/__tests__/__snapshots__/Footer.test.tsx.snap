// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer Component Match Snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "paddingBottom": 13.142857142857142,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#ffffff",
        "borderRadius": 36,
        "borderWidth": 1,
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 16,
        "overflow": "hidden",
      }
    }
  >
    <TextInput
      placeholder="Send Message"
      placeholderTextColor="#ffffff"
      style={
        Object {
          "color": "#ffffff",
          "flex": 1,
          "fontSize": 15.6,
          "height": 73.92857142857143,
          "paddingHorizontal": 24,
          "paddingVertical": 24.642857142857142,
        }
      }
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="footerIcon"
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/icons/send.png",
        }
      }
      style={
        Object {
          "height": 36,
          "marginRight": 16,
          "tintColor": "#ffffff",
          "width": 36,
        }
      }
    />
  </View>
</View>
`;

exports[`Footer Icon Props Match Snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "paddingBottom": 13.142857142857142,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#ffffff",
        "borderRadius": 36,
        "borderWidth": 1,
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 16,
        "overflow": "hidden",
      }
    }
  >
    <TextInput
      placeholder="Send Message"
      placeholderTextColor="#ffffff"
      style={
        Object {
          "color": "#ffffff",
          "flex": 1,
          "fontSize": 15.6,
          "height": 73.92857142857143,
          "paddingHorizontal": 24,
          "paddingVertical": 24.642857142857142,
        }
      }
    />
  </View>
</View>
`;

exports[`Footer TextInput Match Snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "paddingBottom": 13.142857142857142,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#ffffff",
        "borderRadius": 36,
        "borderWidth": 1,
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 16,
        "overflow": "hidden",
      }
    }
  >
    <TextInput
      placeholder="Send Message"
      placeholderTextColor="#ffffff"
      style={
        Object {
          "color": "#ffffff",
          "flex": 1,
          "fontSize": 15.6,
          "height": 73.92857142857143,
          "paddingHorizontal": 24,
          "paddingVertical": 24.642857142857142,
        }
      }
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="footerIcon"
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/assets/icons/send.png",
        }
      }
      style={
        Object {
          "height": 36,
          "marginRight": 16,
          "tintColor": "#ffffff",
          "width": 36,
        }
      }
    />
  </View>
</View>
`;
